README.md
# koop-agol

[![Greenkeeper badge](https://badges.greenkeeper.io/koopjs/koop-provider-agol.svg)](https://greenkeeper.io/)

> ArcGIS Online provider for [Koop](https://github.com/Esri/koop).  Current support for CSV-type items only.  Support for other item types is forth-coming.

[![npm version][npm-img]][npm-url]
[![build status][travis-img]][travis-url]

[npm-img]: https://img.shields.io/npm/v/@koopjs/provider-agol.svg?style=flat-square
[npm-url]: https://www.npmjs.com/package/@koopjs/provider-agol
[travis-img]: https://img.shields.io/travis/koopjs/koop-provider-agol.svg?style=flat-square
[travis-url]: https://travis-ci.org/koopjs/koop-provider-agol

## Install

Koop providers require that you first install Koop. For information on using Koop, see https://koopjs.github.io.

You can add `koop-agol` to your Koop server's dependencies by installing it with npm and adding it to your package.json like so:

```
npm install @koopjs/provider-agol --save
```

## Usage
Koop will register a set of output-plugin endpoints for this provider, e.g., `/agol/:host/:id/FeatureServer`. These routes contain two provider-source route parameters to be aware of:

### `host`
This parameter should be a key representing the AGOL host you want as the source of your item data.  The provider ships with four supported keys: `devext` (https://devext.arcgis.com), `qaext` (https://qaext.arcgis.com), `arcgis` (https://arcgis.com), and finally, `_env`, which can only be use if a `host` value is set the provider's configuration object (see below).

### `id`
This parameter should be the AGOL item id.


## Configure
Koop plugins use the node-config module for setting configuration settings.  See the node-config documentation for usage details.  The Koop AGOL provider current has the following configuration settings:
```json
{
  "koopProviderAgol": {
    "host": "https://my-agol-host",
    "maxCsvSize": 1000000
  }
}
```

### `host`
The `host` configuration setting allows you to match a value to requests that arrive with a `_env` route-parameter.  For example, with configuration file shown above and a request like: `http://localhost:1338/agol/_env/abcdef0123456789abcdef0123456789/FeatureServer`, the provider will look for the item on https://my-agol-host)


### maxCsvSize
The maximum number of bytes a CSV item can be in order to be processed by Koop.  Default value is 5000000.


## Resources

* [Koop Documentation](https://koopjs.github.io)
* [ArcGIS for Developers](http://developers.arcgis.com)
* [ArcGIS REST Documentation](http://resources.arcgis.com/en/help/arcgis-rest-api/)
* [@esri](http://twitter.com/esri)

## Issues

Find a bug or want to request a new feature? Please let us know by submitting an issue.

## Contributing

Esri welcomes contributions from anyone and everyone. Please see our [guidelines for contributing](https://github.com/esri/contributing).

## License

[Apache 2.0](LICENSE)

<!-- [](Esri Tags: ArcGIS Web Mapping GeoJson FeatureServices) -->
<!-- [](Esri Language: JavaScript) -->


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

## [4.1.0] - 2021-04-23
### Added
* Add configuration option to set user-agent in arcgis-rest requests

## [4.0.1] - 2021-04-14
### Changed
* Cast maxCsvSize as a number, it may originate as a ENV variable that is a string

## [4.0.0] - 2020-08-12
### Changed
* Removed all file caching code; limits provider to fetching and translating AGOL CSV items to GeoJSON.

## [3.14.8] - 2017-09-12
### Fixed
* Only use info.name when info is passed

## [3.14.7] - 2017-09-12
### Changed
* Pass csv name in metadata.name

## [3.14.6] - 2017-04-18
### Changed
* Use @koopjs/filesystem-s3 instead of koop-s3fs

## [3.14.5] - 2017-03-27
### Fixed
* Fix field inclusion in export options

## [3.14.4] - 2017-02-27
### Fixed
* Republish

## [3.14.3] - 2017-02-27
### Changed
* Delegate orderByFields, resultOffset and resultRecordCount to upstream featureserver code

## [3.14.2] - 2017-01-31
### Fixed
* Dont look for stat.Metadata when using localfs

## [3.14.1] - 2017-01-31
### Fixed
* Worker.js is compatible with latest koop

## [3.14.0] - 2017-01-31
### Changed
* Compatible with latest Koop 3.x

### Fixed
* Return error when csv has no features
* Set fail when feature service has no features

## [3.13.7] - 2017-01-12
### Fixed
* Be more specific about which cache errors to trap

## [3.13.6] - 2017-01-12
### Changed
* Use different strategy for locking csvs

## [3.13.5] - 2017-01-12
### Fixed
* Intermittent 500 when updating csvs

## [3.13.4] - 2017-01-12
### Fixed
* Fix drop and recache flow

## [3.13.3] - 2017-01-12
### Fixed
* UpdateIfExpired properly updates csv

## [3.13.2] - 2016-12-30
### Changed
* Enqueue exports is called with the result of an import

## [3.13.1] - 2016-12-29
### Fixed
* Koop-Queue dep of worker is added to package.json
* Don't enqueue exports while import is processing

## [3.13.0] - 2016-10-31
### Added
* Sort order sub param to dataset search e.g. (retrieved:ASC, retrieved:DESC)

### Changed
* Refactor export functions into separate model
* Always update exports if lastEditDate changes
* Enqueue exports directly instead of using HTTP after import
* Wait until correct imported file is ready before enqueueing exports

### Fixed
* Remove possible exception in datasets serialization
* Don't allow updates to export job status after finish or fail

## [3.12.2] - 2016-10-20
### Fixed
* Clean up listeners after import service completes

## [3.12.1] - 2016-10-19
### Changed
* Set host before handling POST to datasets controller

## [3.12.0] - 2016-10-15
### Added
* Download from the datasets API
* Add offset parameter to datasets query

### Fixed
* Expire cache when editing info is added for the first time

## [3.11.4] - 2016-09-27
### Fixed
* Worker module references are now correct

## [3.11.3] - 2016-09-23
### Changed
* Adjust generated sql for dataset queries

### Fixed
* Serialization of multiple datasets

## [3.11.2] - 2016-09-22
### Fixed
* Remove es6 property shorthand

## [3.11.1] - 2016-09-22
### Fixed
* Remove arrow function

## [3.11.0] - 2016-09-22
### Changed
* Update keys on dataset API

## [3.10.0] - 2016-09-22
### Added
* Datasets: delete

### Fixed
* Datasets: export
* Removed suspected memory leak with export jobs

## [3.9.0] - 2016-09-21
### Added
* Datasets API

## [3.8.1] - 2016-09-06
### Fixed
* Only call cleanWorkers once and hour
* cleanWorkers is initialized properly

## [3.8.0] - 2016-09-06
### Added
* `DELETE queue/worker/:id` removes a worker from the queue
* `DELETE queue/workers:pattern` removes workers matching `:pattern` from the queue

### Changed
* Do not push jobs into the failed queue on well-handled failures
* `clearFailed` is no longer run every 24 hours

## [3.7.4] - 2016-07-20
### Fixed
* Dont check if csv was updated on first request
* First return of csv feature service has correct name
* Corrected broken spec

## [3.7.3] - 2016-07-20
### Changed
* Update default file ttl to 0 seconds

### Fixed
* Don't return 500 if csv data is not in the cache

## [3.7.2] - 2016-07-06
### Changed
* createExportOptions now takes fields array

## [3.7.1] - 2016-06-08
### Changed
* Worker can use koop-localfs

## [3.7.0] - 2016-05-27
### Added
* Can set default portal from an env. Call `koop/agol/_env` to access it.

### Changed
* Refactor controller into may sub controllers

## [3.6.4] - 2016-05-10
### Fixed
* Better error handling for csv errors

### Changed
* esri-to-geojsonV1.0.2

## [3.6.3] - 2016-05-09
### Fixed
* Only redirect to NGINX when the correct content-type is set

## [3.6.2] - 2016-04-27
### Fixed
* Don't gunzip data for clients that accept gzip when streaming
* Only redirect downloads when the file on S3 has proper headers

## [3.6.1] - 2016-04-25
### Changed
* Specify number of retries for export jobs

## [3.6.0] - 2016-04-25
### Added
* Redirect download traffic to S3 when NGINX is configured

### Changed
* Wait for upload finish event before ending import job
* Do expiration checks after files are served

## [3.5.19] - 2016-04-19
### Changed
* Integrated esri-to-geojson

## [3.5.18] - 2016-04-16
### Changed
* Use `rmdirp` for deleting files

## [3.5.17] - 2016-04-14
### Changed
* Update worker for s3fs

## [3.5.16] - 2016-04-14
### Changed
* Integrated Koop's new fs api

### Fixed
* Always coerce file_min_ttl to integer

## [3.5.15] - 2016-04-13
### Fixed
* Reverse default of `file_min_ttl`

## [3.5.14] - 2016-04-13
### Changed
* Make fudge factor for determining whether a file is out of data configurable

## [3.5.13] - 2016-04-07
### Added
* Log remaining pages left in import job

### Changed
* Revert back to pageQueue drain for ending import job

### Fixed
* Ensure pageCount is always decremented in import service

## [3.5.12] - 2016-04-06
### Fixed
* Protect against case where service has no pages

## [3.5.11] - 2016-04-06
### Fixed

models/cache.js
var Utils = require('../lib/utils.js')
var esriToGeojson = require('esri-to-geojson').fromEsri
var async = require('async')
var CACHE_LIFE = 24 * 60 * 60 * 1000
var path = require('path')
var Portal = require('./portal')

function Cache (options) {
  this._cache = options.cache
  this.files = options.files
  this.log = options.log
  this.featureQueue = options.featureQueue
  this.csvQueue = options.csvQueue
  this.portal = new Portal({log: this.log})
  this.indexFields = options.indexFields
  // max csv file size
  this.maxSize = 5000000
}

/**
 * Wraps Cache.getInfo
 *
 * @param {string} key - a table name in the db
 * @param {function} callback - the callback for when all is gone
 */
Cache.prototype.getInfo = function (key, callback) {
  this._cache.getInfo(key, callback)
}

/**
 * Wraps Cache.updateInfo
 *
 * @param {string} key - a table name in the db
 * @param {object} info - the info to store in the table
 * @param {function} callback  - calls back with err or nothing
 */
Cache.prototype.updateInfo = function (key, info, callback) {
  this._cache.updateInfo(key, info, callback)
}

/**
 * Get the count of the features in the cache
 *
 * @param {string} key - a table name in the db
 * @param {object} options - optional params from req.query (the querystring)
 * @param {function} callback - the callback for when all is gone
 */
Cache.prototype.getCount = function (key, options, callback) {
  this._cache.getCount(key, options, callback)
}

/**
 * Add indexes to a table
 *
 * @param {string} key - the table key in the db
 * @param {object} options - which indexes to add
 * @param {function} callback - calls back after the indexes are added
 */
Cache.prototype.addIndexes = function (key, options, callback) {
  this._cache.addIndexes(key, options, callback)
}

/**
 * Sets a resource in a failed state
 *
 * @param {string} key - a table in the db
 * @param {object} error - an error payload to save in the db
 * @param {function} callback - calls back with an error or nothing
 */
Cache.prototype.setFail = function (key, error, callback) {
  // if (!error) return callback(new Error('cannot set a failure without error info'))
  var self = this
  self._cache.getInfo(key, function (err, info) {
    if (err) return callback(err)
    info = info || {}
    // only set fail if this is not in s3 already
    if (info.status !== 'Cached') info.status = 'Failed'
    info.failedLastImport = true
    info.retrieved = Date.now()
    error.body = error.body || {}
    info.error = {
      timestamp: error.timestamp,
      message: error.msg || error.message,
      request: error.url,
      response: error.body.message,
      code: error.code || error.body.code || 500
    }
    self._cache.updateInfo(key, info, function (err) {
      if (err) return callback(err)
      self.log.info(key + ' set as failed in the DB')
    })
  })
}

/**
 * Create a new table and info and optionally insert rows
 *
 * @param {string} item - a table name in the db
 * @param {number} layer - the layer number of the data
 * @param {object} geojson - data to insert into a new table
 * @param {function} callback - the callback for when the data is inserted
 */
Cache.prototype.insert = function (item, layer, geojson, callback) {
  this._cache.insert('agol', item, geojson, layer, callback)
}

/**
 * Insert rows into an existing table
 *
 * @param {string} item - a table name in the db
 * @param {number} layer - the layer number of the data
 * @param {object} geojson - data to insert into an existing table
 * @param {function} callback - the callback for when the data is inserted
 */
Cache.prototype.insertPartial = function (item, layer, geojson, callback) {
  this._cache.insertPartial('agol', item, geojson, layer, callback)
}

/**
 * Get data from the cache
 *
 * @param {string} item - a table name in the db
 * @param {number} layer - the layer number of the data
 * @param {object} query - query params to apply to the data
 * @param {function} callback - calls back with an error or data
 */
Cache.prototype.get = function (item, layer, query, callback) {
  if (typeof query === 'undefined') query = {}
  query.layer = layer || 0
  this._cache.get('agol', item, query, callback)
}

/**
 * Inserts a never-before-seen feature service into the db
 *
 * @param {object} options - options for which feature service to get
 * @param {function} callback - the calls back with an error or status: processing
 */
Cache.prototype.insertFeatureService = function (options, callback) {
  var self = this
  var table = self._prepareInsert(options)
  self.insert(options.item, options.layer, table, function (err) {
    if (err) return callback(err)
    enqueue(self.featureQueue, options)
    callback(null, table)
  })
}

/**
 * Updates an existing feature service
 *
 * @param {object} options - options for which feature service to get
 * @param {function} callback - the calls back with an error or status: processing
 */
Cache.prototype.updateFeatureService = function (options, callback) {
  var self = this
  var key = 'agol:' + options.item + ':' + options.layer
  self.log.debug('Updating feature service', options)
  self.getInfo(key, function (err, info) {
    if (err) return callback(err)
    // Don't reenquee if the job is already started
    if ((Date.now() - info.importEnqueued) < (12 * 60 * 60 * 1000)) return callback(null, info)
    info.status = 'Expired'
    info.importEnqueued = Date.now()
    self.updateInfo(key, info, function (err) {
      if (err) return callback(err)
      enqueue(self.featureQueue, {
        item: options.item,
        layer: options.layer,
        url: info.url,
        id: options.id
      })
      callback(null, info)
    })
  })
}

function enqueue (queue, options) {
  queue.enqueue('agol', 'importService', [{
    server: options.url || options.itemInfo.url,
    item: options.item,
    layer: options.layer,
    hostId: options.id,
    expires: options.expiration || (Date.now() + CACHE_LIFE)
  }])
}

/**
 * Prepares a table object for insertion into the db by calling out to the service and layer
 *
 * @param {object} options - information used to structure the table
 * @param {function} callback - calls back with an error or the table to insert
 * @private
 */
Cache.prototype._prepareInsert = function (options) {
  var table = {
    status: 'Processing',
    retrieved_at: Date.now(),
    features: [],
    host: {
      id: options.hostId
    },
    // include more in this array when PG-Cache can handle it
    info: {
      itemTitle: options.itemInfo.title,
      version: 3,
      url: options.itemInfo.url,
      type: 'Feature Service',
      metadata: options.itemInfo.metadata,
      itemModified: options.itemInfo.modified,
      _indexFields: false,
      _indexGeometry: false,
      _indexGeohash: false
    }
  }
  return table
}

/**
 * Gets CSV item data from the server or cache, always returns with an error or data
 *
 * @param {object} options - options for which data to get
 * @param {function} callback - calls back with an error or data
 */
Cache.prototype.csv = function (options, callback) {
  var self = this
  var task = {}
  var itemInfo = options.itemInfo
  itemInfo.description = itemInfo.description && itemInfo.description.replace(/\0/g, '')
  itemInfo.name = itemInfo.name && itemInfo.name.replace(/\0/g, '')
  var id = itemInfo.id
  var hostId = options.hostId

  this.get(options.item, options.layer, options.query, function (err, entry) {
    if (err) {
      self.log.debug(err)
      if (/does not exist/i.test(err.message) || /create table/i.test(err.message)) {
        task.passthru = true
      } else if (!/resource not found/i.test(err.message)) {
        err.code = 500
        return callback(err)
      }
    }
    if (entry && entry[0]) return callback(null, null, entry[0])
    if (itemInfo.size > self.maxSize) {
      const sizeError = new Error('The requested CSV exceeds the allowable size of ' + self.maxSize + ' bytes')
      sizeError.code = 413
      return callback(sizeError, null)
    }
    task.url = options.host + Utils.agol_path + id + '/data'
    task.itemInfo = itemInfo
    task.id = id
    task.hostId = hostId
    task.options = options
    task.expires_at = options.expiration || (Date.now() + self.cacheLife)
    task.indexFields = self.indexFields
    self.csvQueue.push(task, callback)
    self.log.info('Request for a new CSV: ' + task.url)
  })
}

/**
 * Get a feature collection from the server and inserts it
 * FeatureCollection are simply JSON features requested in one just (not services)
 *
 * @param {object} params - the set of parameters needed to get the feature collection
 * @param {function} callback - calls back with an error, or info and data
 */
Cache.prototype.featureCollection = function (params, callback) {
  var self = this
  var item = params.item
  var itemInfo = params.itemInfo
  var hostId = params.id

  // for large datasets enforce koop's large data limit
  var options = {}
  options.enforce_limit = true

  self._cache.get('agol', item, options, function (err, entry) {
    if (!err) {
      itemInfo.data = entry
      return callback(null, itemInfo)
    }

    var url = params.host + Utils.agol_path + '/' + item + '/data?f=json'
    Utils.req(url, function (err, data) {
      if (err) return callback(err, null)

      var json = JSON.parse(data.body).featureCollection.layers[0].featureSet

      var geojson
      try {
        geojson = esriToGeojson(json)
      } catch (e) {
        return callback(e)
      }

      geojson.name = itemInfo.title
      geojson.updated_at = itemInfo.modified
      geojson.host = {id: hostId}
      geojson.info = {
        type: 'Feature Collection',
        version: 3,
        url: url
      }
      self.insert(item, params.layer, geojson, function (err, success) {
        if (success) {
          itemInfo.data = [geojson]
          return callback(null, itemInfo, geojson)
        }
        callback(err)
      })
    })
  })
}

/**
 * Method to remove all the data in each export dir
 *
 * @param {string} dir - The base dir to remove all files
 * @param {function} callback - The callback.
 * @private
 */
Cache.prototype._removeExportDirs = function (dir, callback) {
  var self = this
  var dirs = ['files/', 'tiles/']
  async.each(dirs, function (base, callback) {
    self.files.rmdirp(path.join(base, dir), function (err) {
      callback(err)
    })
  }, function (err) {
    callback(err)
  })
}

/**
 * Drops the item from the cache
 * will remove all exported files as well,
 * but will only remove "latest" files if forceDelete is set
 * @param {string} itemid - id of the item
 * @param {integer} layer - the layer of data
 * @param {object} options - optional params from req.query (the querystring)
 * @param {function} callback - the callback for when all is gone
 */
Cache.prototype.drop = function (item, layer, options, callback) {
  var self = this
  if (typeof options === 'function') callback = options
  layer = layer || 0
  var dir = [ item, layer ].join('_')
  self._cache.remove('agol', item, {layer: layer}, function (err, res) {
    if (err) return callback(err)
    self._removeExportDirs(dir, function (err, success) {
      if (!options || !options.forceDelete) return callback(err, true)
      self.files.rmdirp('latest/files/' + dir, function (err, res) {
        callback(err)
      })
    })
  })
}

/**
 * Checks to see if an item is expired or not
 *
 * @param {object} info - metadata from the service
 * @param {object} options
 * @param {function} callback - callback to call when the requests are done
 */
Cache.prototype.checkExpiration = function (info, options, callback) {
  if (info.type === 'CSV') return this._csvExpiration(info, options, callback)

  // always send hosted services to hostedFeatureExp
  var hosted = new RegExp(/services[0-9]?\.arcgis/)
  if (info.lastEditDate || hosted.test(info.url)) return this._hostedFeatureExpiration(info, options.layer, callback)

  this._featureExpiration(info, options.layer, callback)
}

/**
 * Checks to see if a csv is expired or not
 *
 * @param {object} info - metadata from the service
 * @param {function} callback - callback to call when the requests are done
 * @private
 */
Cache.prototype._csvExpiration = function (info, options, callback) {
  this.portal.getItem(options.host, options.item, function (err, item) {
    if (err) return callback(err)
    var expired = item.modified > info.retrieved_at
    callback(null, expired, info)
  })
}

/**
 * Checks to see if a feature service is expired or not
 *
 * @param {object} info - metadata from the service
 * @param {number} layerId - the number of the layer in the request
 * @param {function} callback - callback to call when the requests are done
 * @private
 */
Cache.prototype._featureExpiration = function (info, layer, callback) {
  var expired = Date.now() > info.expires_at
  callback(null, expired, info)
}

/**
 * Checks to see if a hosted feature service is expired or not
 *
 * @param {object} info - metadata from the service
 * @param {number} layerId - the number of the layer in the request
 * @param {function} callback - callback to call when the requests are done
 * @private
 */
Cache.prototype._hostedFeatureExpiration = function (info, layer, callback) {
  // if we have lastEditDate on the service we need to check for expiration every time this data is requested
  var self = this
  var start = Date.now()
  Utils.getLayerInfo(info.url, layer, function (err, layerInfo) {
    self.log.debug('Checking hosted feature service expiration took:', (Date.now() - start) / 1000, 'seconds')
    if (err) return callback(err)
    if (!layerInfo.editingInfo) return callback(null, true, info)
    // Handle the case where there was no editing info the first time we imported
    if (layerInfo.editingInfo.lastEditDate && !info.lastEditDate) return callback(null, true, info)
    var expired = layerInfo.editingInfo.lastEditDate > info.lastEditDate
    callback(null, expired, info)
  })
}

module.exports = Cache
